clarity/src/vm/types/signatures.rs

```diff
@@ -1138,19 +1366,35 @@ impl TypeSignature {
                     .ok_or(CheckErrors::InvalidTypeDescription)?;
                 if let SymbolicExpressionType::Atom(ref compound_type) = compound_type.expr {
                     match compound_type.as_ref() {
-                        "list" => TypeSignature::parse_list_type_repr(rest, accounting),
+                        "list" => TypeSignature::parse_list_type_repr(epoch, rest, accounting),
                         "buff" => TypeSignature::parse_buff_type_repr(rest),
                         "string-utf8" => TypeSignature::parse_string_utf8_type_repr(rest),
                         "string-ascii" => TypeSignature::parse_string_ascii_type_repr(rest),
-                        "tuple" => TypeSignature::parse_tuple_type_repr(rest, accounting),
-                        "optional" => TypeSignature::parse_optional_type_repr(rest, accounting),
-                        "response" => TypeSignature::parse_response_type_repr(rest, accounting),
+                        "tuple" => TypeSignature::parse_tuple_type_repr(epoch, rest, accounting),
+                        "optional" => {
+                            TypeSignature::parse_optional_type_repr(epoch, rest, accounting)
+                        }
+                        "response" => {
+                            TypeSignature::parse_response_type_repr(epoch, rest, accounting)
+                        }
                         _ => Err(CheckErrors::InvalidTypeDescription),
                     }
                 } else {
                     Err(CheckErrors::InvalidTypeDescription)
                 }
             }
+            SymbolicExpressionType::TraitReference(_, ref trait_definition)
+                if epoch < StacksEpochId::Epoch21 =>
+            {
+                match trait_definition {
+                    TraitDefinition::Defined(trait_id) => {
+                        Ok(TypeSignature::TraitReferenceType(trait_id.clone()))
+                    }
+                    TraitDefinition::Imported(trait_id) => {
+                        Ok(TypeSignature::TraitReferenceType(trait_id.clone()))
+                    }
+                }
```

Comment:
What happens if `epoch == StacksEpochId::Epoch21`?  Is that reachable?

Reply:
Yes, it will fall through to the next case of the match, where a `CallableType` is used instead of a `TraitReferenceType`.